## **Python and Big Data Analytics Course Project**

## *Agricultural Production Analysis*

<img src="https://cdn.pixabay.com/photo/2014/12/11/02/55/cereals-563796__340.jpg" width=1200  height=450/>

Kristine Dege

Grazina

##### **The Dataset used in the Course Project is downloaded from the website https://data.world/agriculture/**
##### Crop statistics for 173 products in Africa, the Americas, Asia, Europe, and Oceania (in the Course Project Eastern Europe Data will be used only).
##### Crops Primary, Fibre Crops Crop statistics are recorded for 173 products, covering the following categories: Crops Primary, Fibre Crops Primary, Cereals, Coarse Grain, Citrus Fruit, Fruit, Jute & Jute-like Fibres, Oilcakes Equivalent, Oil crops Primary, Pulses, Roots and Tubers, Treenuts and Vegetables and Melons. Data are expressed in terms of area harvested, production quantity, yield and seed quantity. The objective is to comprehensively cover production of all primary crops for all countries and regions in the world.
##### Data source: The Food and Agriculture Organization of the United Nations (FAO).

##### The objective of the Project is to ..........!!!!!!

*The full path used to get the dataset:*
* browser = webdriver.Edge('C:/Users/Kristine/Drivers/msedgedriver.exe')
* time.sleep(2)
* url = 'https://data.world/agriculture/'
* browser.get(url)
* dataset_button = browser.find_element(By.XPATH, '//*[@id="profile_view.main_tabs-pane-overview"]/div/div/div[1]/div/ul/li[2]/div').click() *# Click on Datasets button.*
* crop_pr_but = browser.find_element(By.XPATH, '//*[@id="profile_view.main_tabs-pane-resources"]/div/div/div[2]/div[2]/div/div/div[9]/div/div[2]/div[1]/a').click() *# Click Crop_Production datasets button.  Overview and discription of datasets are available.*
* crop_ds = browser.find_element(By.XPATH, '//*[@id="dataset_view.main_tabs-pane-details"]/div/div/div[1]/div[4]/div[8]/div/div/div[1]/div/div[1]/a[2]/span').click() *# In case you are not a registered user then sign-up model will be opened. There is an option to proceed with Google/ Facebook/ GitHub/ Twitter account.*

* browser.find_element(By.XPATH, '//*[@id="create_user.email"]').send_keys(email)  
* browser.find_element(By.XPATH, '//*[@id="create_user.displayName"]').send_keys(Name)
* browser.find_element(By.XPATH, '//*[@id="create_user.agentid"]').send_keys(Username)
* browser.find_element(By.XPATH, '//*[@id="create_user.password1"]').send_keys(Password)
* browser.find_element(By.XPATH, '/html/body/div[13]/div[2]/div/div/div/div/div[2]/div/div/div/form/button').click() *# Click on Continue button.*
* browser.find_element(By.XPATH, '/html/body/div[2]/div[2]/div/div/div/div/div[2]/a').click() *# Click on button Create a Project.*
* browser.find_element(By.XPATH, '//*[@id="root"]/div/div[2]/div[2]/div/div[1]/div[2]/div/div/div[4]/div/div/div/div/div[2]/div[1]/a').click() *# Now you should verify your e-mail address and then click on Crop Production dataset.*
* browser.find_element(By.XPATH, '//*[@id="dataset_view.main_tabs-pane-details"]/div/div/div[1]/div[4]/div[8]/div/div/div[1]/div/div[1]/a[2]').click() *# Click on dataset Crop Production in Eastern Europe.*
* browser.find_element(By.XPATH, '//*[@id="file_renderer.modal_common.export_dropdown"]/div/span[2]').click() *# Click on Download drop-down menu.*
* browser.find_element(By.XPATH, '/html/body/div[2]/div/ul/span[1]/li/a/div[2]/div[1]').click() *# Click on Share URL. Now you can copy the path to download csv file from web-site.*

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
from collections import Counter
import seaborn as sns
import geopandas as gpd
import geocoder
import requests
from geopy.geocoders import Nominatim
import pycountry_convert as pc
import numpy as np

In [2]:
df = pd.read_csv('https://query.data.world/s/zmbllbphuegmiqcjpgtnrl3xew2jxh')

In [3]:
df.sample(5)

Area Code         Area  Item Code                              Item  \
4328        104      Ireland        401       Chillies and peppers, green   
8820        203        Spain        711  Anise, badian, fennel, coriander   
9404        211  Switzerland        515                            Apples   
2994         67      Finland        463             Vegetables, fresh nes   
5102        126    Lithuania        426               Carrots and turnips   

      Element Code     Element    Unit     Y1961 Y1961F      Y1962  ...  \
4328          5510  Production  tonnes       NaN    NaN        NaN  ...   
8820          5419       Yield   hg/ha       NaN    NaN        NaN  ...   
9404          5419       Yield   hg/ha  999259.0     Fc  1000833.0  ...   
2994          5510  Production  tonnes   19400.0    NaN    11300.0  ...   
5102          5510  Production  tonnes       NaN    NaN        NaN  ...   

         Y2015  Y2015F     Y2016  Y2016F     Y2017  Y2017F     Y2018  Y2018F  \
4328     288.0      Im     357.0      Im     340.0      Im       NaN     NaN   
8820    6226.0      Fc    6291.0      Fc    7153.0      Fc       NaN     NaN   
9404  554621.0      Fc  589777.0      Fc  383248.0      Fc  783883.0      Fc   
2994   12231.0     NaN   12269.0     NaN   12431.0     NaN       0.0     NaN   
5102   43751.0     NaN   49315.0     NaN   41902.0     NaN   48380.0     NaN   

         Y2019  Y2019F  
4328       NaN     NaN  
8820       NaN     NaN  
9404  507832.0      Fc  
2994       0.0     NaN  
5102   45300.0     NaN  

[5 rows x 125 columns]

In [15]:
df.shape

(10557, 125)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10557 entries, 0 to 10556
Columns: 125 entries, Area Code to Y2019F
dtypes: float64(59), int64(3), object(63)
memory usage: 10.1+ MB


In [29]:
# Deleting all forecast columns as they are empty

for col in df.columns:
    if 'F' in col:
        del df[col]
   
df.head(3)

Area Code     Area  Item Code    Item  Element Code         Element  \
0          3  Albania        515  Apples          5312  Area harvested   
1          3  Albania        515  Apples          5419           Yield   
2          3  Albania        515  Apples          5510      Production   

     Unit    Y1961   Y1962   Y1963  ...     Y2010     Y2011     Y2012  \
0      ha      NaN     NaN     NaN  ...    3913.0    3450.0    3719.0   
1   hg/ha      NaN     NaN     NaN  ...  139545.0  185507.0  191718.0   
2  tonnes  10004.0  8039.0  7931.0  ...   54604.0   64000.0   71300.0   

      Y2013     Y2014     Y2015     Y2016     Y2017     Y2018     Y2019  
0    3838.0    3863.0    4008.0    4230.0    4346.0    4294.0    4407.0  
1  197316.0  212426.0  228882.0  240028.0  221671.0  252387.0  240374.0  
2   75730.0   82060.0   91736.0  101532.0   96338.0  108375.0  105933.0  

[3 rows x 66 columns]

In [30]:
#Changing years column name from'Y2010' to numeric only

df.columns = map(lambda x : x.replace("Y", ""), df.columns)

Only the first seven columns do not have null values. The reasons of null values are geographical area changes, climatic conditions in different areas, and vast product range. So, it is resonable to drop empty rows.

In [31]:
df.dropna(inplace=True)
df.shape

(2672, 66)

By deleting empty forecast columns we have reduced the total number of columns from 125 to 66.

By deleting empty rows we have reduced the total number of rows from 10557 to 2672.

In [6]:
df['Item'].value_counts()

Roots and Tubers, Total    69
Potatoes                   69
Vegetables Primary         65
Tomatoes                   64
Wheat                      62
                           ..
Beans, dry                  3
Vetches                     3
Gooseberries                3
Blueberries                 1
Quinces                     1
Name: Item, Length: 95, dtype: int64

##### The most cultivated agriculture products are: roots, potatoes, barley, primary vegetables and cabbages.
##### The less cultivated products in Europe are: tung nuts, green coffee, pepper, mangoes.

In [32]:
# Converting country name (column 'Area') into continent name

def convert(row):
    try:
        cn_code = pc.country_name_to_country_alpha2(row.Area, cn_name_format='default')
        conti_code = pc.country_alpha2_to_continent_code(cn_code)
        return conti_code
    except:
        None # If can't find such country then return None

In [33]:
# Creating nem column 'Continent'

df['Continent'] = df.apply(convert, axis = 1)
df.sample(3)

Area Code     Area  Item Code            Item  Element Code     Element  \
7520        183  Romania        267  Sunflower seed          5510  Production   
6649        173   Poland        187       Peas, dry          5510  Production   
156           3  Albania        157      Sugar beet          5419       Yield   

        Unit      1961      1962      1963  ...       2011       2012  \
7520  tonnes  481400.0  449500.0  505900.0  ...  1789326.0  1398203.0   
6649  tonnes   54000.0   63000.0   64000.0  ...    36893.0    44978.0   
156    hg/ha  146531.0  146489.0  156977.0  ...   200000.0   200000.0   

           2013       2014       2015       2016       2017       2018  \
7520  2142087.0  2189309.0  1785771.0  2032340.0  2912743.0  3062690.0   
6649    33232.0    44421.0    91254.0    95161.0    82768.0    25740.0   
156    200000.0   226775.0   270578.0   305264.0   341721.0   391524.0   

           2019  Continent  
7520  3569150.0         EU  
6649    30880.0         EU  
156    381429.0         EU  

[3 rows x 67 columns]

In [34]:
df[df['Continent']==None] ## should be None values???!!!


Empty DataFrame
Columns: [Area Code, Area, Item Code, Item, Element Code, Element, Unit, 1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, Continent]
Index: []

[0 rows x 67 columns]

In [35]:
# Converting conutry name into short name

def country(row):
    try:
        countr_code =  pc.country_name_to_country_alpha2(row.Area, cn_name_format='default')
        return countr_code
    except:
        return None

In [36]:
#Creating new column with country short name

df['Country'] = df.apply(country, axis = 1)
df.sample(3)

Area Code                                               Area  Item Code  \
2407         54                                            Denmark        397   
9933        229  United Kingdom of Great Britain and Northern I...        550   
3713         84                                             Greece        406   

                        Item  Element Code     Element    Unit      1961  \
2407  Cucumbers and gherkins          5419       Yield   hg/ha  211826.0   
9933                Currants          5510  Production  tonnes   21438.0   
3713                  Garlic          5419       Yield   hg/ha   28458.0   

          1962      1963  ...       2012       2013       2014       2015  \
2407  169444.0  232135.0  ...  1739727.0  2362651.0  2840000.0  2501250.0   
9933   23674.0   32006.0  ...    11910.0    16986.0    12698.0    15224.0   
3713   33298.0   35887.0  ...    64312.0    69495.0    81700.0    82729.0   

           2016       2017       2018       2019  Continent  Country  
2407  2474000.0  2215479.0  2362857.0  2753333.0         EU       DK  
9933    12238.0    13899.0    15141.0    13448.0         EU       GB  
3713    78579.0    81848.0    78590.0    75690.0         EU       GR  

[3 rows x 68 columns]

In [37]:
#Creating latitude and logitude values

geolocator = Nominatim(user_agent='berzina_e@yahoo.com')
def geolocate(country):
    try:
        # Geolocate the center of the country
        loc = geolocator.geocode(country.Country)
        # And return latitude and longitude
        return loc.latitude, loc.longitude
    except:
        # Return missing value as some countries as USSR do not exist any more
        return None

In [38]:
#Creating new column Geolocatin with latitude and longitude values

df['GeoLocation'] = df.apply(geolocate, axis = 1)
df.sample(3)

In [17]:
#Creating two new columns

df[['Latitude', 'Longitude']] = df['GeoLocation'].astype(str).str.strip('()').str.split(', ', expand=True).rename(columns={0:'Latitude', 1:'Longitude'}) 

In [18]:
df.sample(3)

Area Code     Area  Item Code    Item  Element Code         Element  \
0          3  Albania        515  Apples          5312  Area harvested   
1          3  Albania        515  Apples          5419           Yield   
2          3  Albania        515  Apples          5510      Production   

     Unit     1961    1962    1963  ...      2015      2016      2017  \
0      ha      NaN     NaN     NaN  ...    4008.0    4230.0    4346.0   
1   hg/ha      NaN     NaN     NaN  ...  228882.0  240028.0  221671.0   
2  tonnes  10004.0  8039.0  7931.0  ...   91736.0  101532.0   96338.0   

       2018      2019  Continent  Country                GeoLocation  \
0    4294.0    4407.0         EU       AL  (33.2588817, -86.8295337)   
1  252387.0  240374.0         EU       AL  (33.2588817, -86.8295337)   
2  108375.0  105933.0         EU       AL  (33.2588817, -86.8295337)   

     Latitude    Longitude  
0  33.2588817  -86.8295337  
1  33.2588817  -86.8295337  
2  33.2588817  -86.8295337  

[3 rows x 71 columns]

In [19]:
import plotly.express as px

In [20]:
fig = px.scatter_geo(data, lat='Latitude', lon='Longitude')
fig.update_layout(title = 'Agricultural production Areas in Europe', title_x=0.5)
fig.show()

In [ ]:
n = 10
df.nlargest(n, "Col_Name", keep = "all")

In [ ]:
n=10
df.sort_values("Col_Name", ascending=False).head(n)

In [130]:
# Choosing rows with Latvian area

df_lv = df.loc[df['Area'] =='Latvia'] 

In [131]:
#Deleting empty years (Y1961-Y1991) as Latvia was a part of USSR

df_lv.drop(df.iloc[:, 7:38], inplace=True, axis=1)
df_lv.head(4)

C:\Users\Д\AppData\Local\Temp\ipykernel_636\317519920.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_lv.drop(df.iloc[:, 7:38], inplace=True, axis=1)


Area Code    Area  Item Code                 Item  Element Code  \
4819        119  Latvia        221  Almonds, with shell          5312   
4820        119  Latvia        221  Almonds, with shell          5510   
4821        119  Latvia        515               Apples          5312   
4822        119  Latvia        515               Apples          5419   

             Element    Unit     1992     1993     1994  ...     2010  \
4819  Area harvested      ha      NaN      NaN      NaN  ...      NaN   
4820      Production  tonnes      NaN      NaN      NaN  ...      NaN   
4821  Area harvested      ha  12000.0  12000.0  12000.0  ...   3257.0   
4822           Yield   hg/ha  42500.0  77500.0  15833.0  ...  31701.0   

         2011     2012     2013     2014     2015     2016     2017     2018  \
4819      NaN      NaN      NaN      NaN      NaN      NaN      NaN      0.0   
4820      NaN      NaN      NaN      NaN      NaN      NaN      NaN      0.0   
4821   3379.0   3136.0   2762.0   2653.0   2417.0   2458.0   3331.0   3200.0   
4822  22199.0  30131.0  53407.0  36272.0  32333.0  40090.0  22408.0  43438.0   

         2019  
4819      0.0  
4820      0.0  
4821   3440.0  
4822  30203.0  

[4 rows x 35 columns]

In [132]:
# As geografical area included in dataset is very wide, there are agricultural products that are not produced in Latvia.

df_lv.isna().sum()

Area Code         0
Area              0
Item Code         0
Item              0
Element Code      0
Element           0
Unit              0
1992            111
1993            111
1994            111
1995            111
1996            108
1997             94
1998             95
1999             86
2000             83
2001             83
2002             83
2003             83
2004             80
2005             80
2006             77
2007             77
2008             80
2009             79
2010             79
2011             79
2012             79
2013             79
2014             79
2015             79
2016             79
2017             79
2018             51
2019             51
dtype: int64

In [138]:
df_lv.T.head()

4821    4822    4823    4830    4831    4832         4836  \
Area Code        119     119     119     119     119     119          119   
Area          Latvia  Latvia  Latvia  Latvia  Latvia  Latvia       Latvia   
Item Code        515     515     515      44      44      44          558   
Item          Apples  Apples  Apples  Barley  Barley  Barley  Berries nes   
Element Code    5312    5419    5510    5312    5419    5510         5312   

                     4837         4838                          4848  ...  \
Area Code             119          119                           119  ...   
Area               Latvia       Latvia                        Latvia  ...   
Item Code             558          558                           358  ...   
Item          Berries nes  Berries nes  Cabbages and other brassicas  ...   
Element Code         5419         5510                          5312  ...   

                       5043           5044                     5045  \
Area Code               119            119                      119   
Area                 Latvia         Latvia                   Latvia   
Item Code              1726           1726                     1720   
Item          Pulses, Total  Pulses, Total  Roots and Tubers, Total   
Element Code           5419           5510                     5312   

                                 5046                     5047  \
Area Code                         119                      119   
Area                           Latvia                   Latvia   
Item Code                        1720                     1720   
Item          Roots and Tubers, Total  Roots and Tubers, Total   
Element Code                     5419                     5510   

                             5048                 5050                5053  \
Area Code                     119                  119                 119   
Area                       Latvia               Latvia              Latvia   
Item Code                    1723                 1723                1735   
Item          Sugar Crops Primary  Sugar Crops Primary  Vegetables Primary   
Element Code                 5312                 5510                5312   

                            5054                5055  
Area Code                    119                 119  
Area                      Latvia              Latvia  
Item Code                   1735                1735  
Item          Vegetables Primary  Vegetables Primary  
Element Code                5419                5510  

[5 rows x 93 columns]

In [137]:
df_lv.T.T.head()

Area Code    Area Item Code    Item Element Code         Element    Unit  \
4821       119  Latvia       515  Apples         5312  Area harvested      ha   
4822       119  Latvia       515  Apples         5419           Yield   hg/ha   
4823       119  Latvia       515  Apples         5510      Production  tonnes   
4830       119  Latvia        44  Barley         5312  Area harvested      ha   
4831       119  Latvia        44  Barley         5419           Yield   hg/ha   

          1992      1993      1994  ...      2010     2011     2012     2013  \
4821   12000.0   12000.0   12000.0  ...    3257.0   3379.0   3136.0   2762.0   
4822   42500.0   77500.0   15833.0  ...   31701.0  22199.0  30131.0  53407.0   
4823   51000.0   93000.0   19000.0  ...   10325.0   7501.0   9449.0  14751.0   
4830  350400.0  275300.0  266500.0  ...  100400.0  95900.0  86300.0  84400.0   
4831   12372.0   16546.0   18053.0  ...   22749.0  24682.0  28806.0  27559.0   

          2014     2015     2016     2017      2018     2019  
4821    2653.0   2417.0   2458.0   3331.0    3200.0   3440.0  
4822   36272.0  32333.0  40090.0  22408.0   43438.0  30203.0  
4823    9623.0   7815.0   9854.0   7464.0   13900.0  10390.0  
4830  118200.0  99300.0  94400.0  70300.0  118300.0  86800.0  
4831   35431.0  38781.0  30000.0  34267.0   25892.0  35184.0  

[5 rows x 35 columns]

In [13]:
df.loc[df['Area'] =='Lithuania']

Area Code       Area  Item Code                              Item  \
5056        126  Lithuania        221               Almonds, with shell   
5057        126  Lithuania        221               Almonds, with shell   
5058        126  Lithuania        711  Anise, badian, fennel, coriander   
5059        126  Lithuania        711  Anise, badian, fennel, coriander   
5060        126  Lithuania        711  Anise, badian, fennel, coriander   
...         ...        ...        ...                               ...   
5309        126  Lithuania       1729                   Treenuts, Total   
5310        126  Lithuania       1729                   Treenuts, Total   
5311        126  Lithuania       1735                Vegetables Primary   
5312        126  Lithuania       1735                Vegetables Primary   
5313        126  Lithuania       1735                Vegetables Primary   

      Element Code         Element    Unit  Y1961  Y1962  Y1963  ...  \
5056          5312  Area harvested      ha    NaN    NaN    NaN  ...   
5057          5510      Production  tonnes    NaN    NaN    NaN  ...   
5058          5312  Area harvested      ha    NaN    NaN    NaN  ...   
5059          5419           Yield   hg/ha    NaN    NaN    NaN  ...   
5060          5510      Production  tonnes    NaN    NaN    NaN  ...   
...            ...             ...     ...    ...    ...    ...  ...   
5309          5312  Area harvested      ha    NaN    NaN    NaN  ...   
5310          5510      Production  tonnes    NaN    NaN    NaN  ...   
5311          5312  Area harvested      ha    NaN    NaN    NaN  ...   
5312          5419           Yield   hg/ha    NaN    NaN    NaN  ...   
5313          5510      Production  tonnes    NaN    NaN    NaN  ...   

         Y2010     Y2011     Y2012     Y2013     Y2014     Y2015     Y2016  \
5056       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
5057       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
5058    7400.0   25600.0    5400.0    5600.0    6400.0    5514.0    6548.0   
5059    1239.0     859.0    5370.0    5000.0    3906.0    4971.0   11069.0   
5060     917.0    2200.0    2900.0    2800.0    2500.0    2741.0    7248.0   
...        ...       ...       ...       ...       ...       ...       ...   
5309       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
5310       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
5311   12300.0   12646.0   11160.0   11323.0   10745.0   10472.0   13757.0   
5312  140904.0  218974.0  234794.0  194635.0  218659.0  181717.0  186829.0   
5313  173312.0  276915.0  262030.0  220385.0  234949.0  190294.0  257021.0   

         Y2017     Y2018     Y2019  
5056       NaN       0.0       0.0  
5057       NaN       0.0       0.0  
5058    7012.0       NaN       NaN  
5059    6409.0       NaN       NaN  
5060    4494.0       NaN       NaN  
...        ...       ...       ...  
5309       NaN       0.0       0.0  
5310       NaN       0.0       0.0  
5311   12998.0    8640.0    8900.0  
5312  164220.0  192477.0  199831.0  
5313  213453.0  166300.0  177850.0  

[258 rows x 66 columns]

In [9]:
# The dataset of selling prices of crop products is downloaded from:
#  https://ec.europa.eu/eurostat/databrowser/view/apri_ap_crpouta

Area Code
0          3
1          3
2          3
3          3
4          3